File: dt-spark-core/src/main/scala/com/kevin/scala/PartitionerDemo.scala
package com.kevin.scala

import org.apache.spark._

import scala.collection.mutable.ArrayBuffer

/**
  * 自定义分区器
  */
object PartitionerDemo {

  def main(args: Array[String]): Unit = {
    // 1.创建SparkConf
    val conf = new SparkConf().setAppName("PartitionerDemo").setMaster("local")
    // 2.创建SparkContext
    val sc = new SparkContext(conf)
    // 创建元组集合
    val list = List((1,"a"),(2,"b"),(3,"c"),(4,"d"),(5,"e"),(6,"f"))
    // 3.转为rdd并设置分区为2
    val rdd = sc.parallelize(list,2)
    // 4.自定义分区前
    rdd.mapPartitionsWithIndex((index,iter) => {
      val result = new ArrayBuffer[(Int,String)]()
      while (iter.hasNext) {
        val next = iter.next()
        println("自定义分区前 partitionID = "+index+" , value = "+next)
        result.+=(next)
      }
      println(result.length)
      result.iterator
    }, true).collect()
    println("----------")
    // 5.自定义分区规则
    val partitionBy = rdd.partitionBy(new MyPartitioner(4))

    // 6.自定义分区后
    partitionBy.mapPartitionsWithIndex((index,iter) => {
      val result = new ArrayBuffer[(Int,String)]()
      while (iter.hasNext) {
        val next = iter.next()
        println("自定义分区后 partitionID = "+index+" , value = "+next)
        result.+=(next)
      }
      println(result.length)
      result.iterator
    },true).collect()
    // 7.关闭sc
    sc.stop()

  }
}

/**
  * 自定义分区类，需继承Partitioner类
  * @param num
  */
class MyPartitioner(num: Int) extends Partitioner {
  // 覆盖分区数
  override def numPartitions: Int = num
  // 覆盖分区号获取函数
  override def getPartition(key: Any): Int = {
    if(key.toString.toInt % 2 == 0) {
      0
    } else {
      1
    }
  }
}

File: dt-spark-core/src/main/java/com/kevin/java/PartitionerDemo.java
package com.kevin.java;

import org.apache.spark.Partitioner;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function2;
import scala.Tuple2;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

/**
 * @author kevin
 * @version 1.0
 * @description     自定义分区器
 * @createDate 2019/1/2
 */
public class PartitionerDemo {

    public static void main(String[] args) {

        SparkConf conf = new SparkConf().setAppName("PartitionerDemo").setMaster("local");
        JavaSparkContext sc = new JavaSparkContext(conf);

        // 将Tuple2的数据转为rdd
        JavaPairRDD<Integer, String> rdd = sc.parallelizePairs(Arrays.asList(
                new Tuple2<Integer, String>(1,"java"),
                new Tuple2<Integer, String>(2,"scala"),
                new Tuple2<Integer, String>(3,"python"),
                new Tuple2<Integer, String>(4,"c#"),
                new Tuple2<Integer, String>(5,"c"),
                new Tuple2<Integer, String>(6,"c++")
        ),2);

        // 自定义分区前
        rdd.mapPartitionsWithIndex(new Function2<Integer, Iterator<Tuple2<Integer, String>>, Iterator<String>>() {

            @Override
            public Iterator<String> call(Integer index, Iterator<Tuple2<Integer, String>> iter) throws Exception {
                List<String> list = new ArrayList<String>();
                while(iter.hasNext()){
                    System.out.println("自定义分区前 partitionID = "+index+" , value = "+iter.next());
                }
                return list.iterator();
            }
        },true).collect();

        System.out.println("------");

        JavaPairRDD<Integer, String> partitionRDD = rdd.partitionBy(new Partitioner() {
            // 设置分区数
            @Override
            public int numPartitions() {
                return 2;
            }

            // 自定义分区规则，对输入的key做计算，然后返回该key的分区ID，范围一定是0到numPartitions-1
            @Override
            public int getPartition(Object key) {
                int i = (int) key;
                if (i % 2 == 0) {
                    return 0;
                }
                return 1;
            }
        });

        // 自定义分区后
        partitionRDD.mapPartitionsWithIndex(new Function2<Integer, Iterator<Tuple2<Integer, String>>, Iterator<String>>() {

            @Override
            public Iterator<String> call(Integer index, Iterator<Tuple2<Integer, String>> iter) throws Exception {
                List<String> list = new ArrayList<String>();
                while(iter.hasNext()){
                    System.out.println("自定义分区后 partitionID = "+index+" , value = "+iter.next());
                }
                return list.iterator();
            }
        },true).collect();

        sc.close();

    }
}
